File: runescript-commons/src/main/java/me/waliedyassen/runescript/type/primitive/PrimitiveType.java
/*
 * Copyright (c) 2020 Walied K. Yassen, All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package me.waliedyassen.runescript.type.primitive;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.var;
import me.waliedyassen.runescript.type.Type;
import me.waliedyassen.runescript.type.serializer.TypeSerializer;
import me.waliedyassen.runescript.type.stack.StackType;

import java.util.Arrays;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Represents the primitive types within our type system.
 *
 * @author Walied K. Yassen
 */
@RequiredArgsConstructor
public enum PrimitiveType implements Type {

    /**
     * The undefined primitive type.
     */
    UNDEFINED('\ufff0', "undefined", null, null, null),

    /**
     * The hook primitive type.
     */
    HOOK('\ufff1', "hook", null, null, null),

    /**
     * The void primitive type.
     */
    VOID('\ufff2', "void", null, null, null),

    /**
     * The byte special type.
     */
    BYTE('\ufff3', "byte", null, null, TypeSerializer.BYTE),

    /**
     * The short special type.
     */
    SHORT('\ufff4', "short", null, null, TypeSerializer.SHORT),

    /**
     * The tribyte special type.
     */
    TRIBYTE('\ufff5', "tribyte", null, null, TypeSerializer.TRIBYTE),

    /**
     * The type special type.
     */
    TYPE('\ufff6', "type", null, null, TypeSerializer.TYPE),

    /**
     * The param configuration type.
     */
    PARAM('\uffd0', "param", null, null, null),

    /**
     * The floor overlay configuration type.
     */
    FLO('\uffd1', "flo", null, null, null),

    /**
     * The floor underlay configuration type.
     */
    FLU('\uffd2', "flu", null, null, null),

    /**
     * The var configuration type.
     */
    VAR('\uffd3', "var", null, null, null),
    /**
     * The var configuration type.
     */
    VARBIT('\uffd4', "varbit", null, null, null),

    /**
     * The var client string configuration type.
     */
    VARCSTR('\uffd5', "varcstr", null, null, null),

    /**
     * The var client integer configuration type.
     */
    VARCINT('\uffd6', "varcint", null, null, null),

    /**
     * The null special type.
     */
    NULL('\uffd7', null, null, null, null),

    // All the types below are verified to be part of script var type.

    /**
     * The integer primitive type.
     */
    INT('i', "int", StackType.INT, 0, TypeSerializer.INT),

    /**
     * The string primitive type.
     */
    STRING('s', "string", StackType.STRING, "", TypeSerializer.STRING),

    /**
     * The spotanim primitive type.
     */
    SPOTANIM('t', "spotanim", StackType.INT, -1, TypeSerializer.INT),

    /**
     * The seq primitive type.
     */
    SEQ('A', "seq", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The stat primitive type.
     */
    STAT('S', "stat", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The synth primitive type.
     */
    SYNTH('P', "synth", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The coordgrid primitive type.
     */
    COORDGRID('c', "coordgrid", StackType.INT, -1, TypeSerializer.INT),

    /**
     * The char primitive type.
     */
    CHAR('z', "char", StackType.INT, -1, TypeSerializer.BYTE),

    /**
     * The fontmetrics primitive type.
     */
    FONTMETRICS('f', "fontmetrics", StackType.INT, -1, TypeSerializer.INT),

    /**
     * The maparea primitive type.
     */
    MAPAREA('`', "maparea", StackType.INT, -1, TypeSerializer.INT),

    /**
     * The enum primitive type.
     */
    ENUM('g', "enum", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The npc primitive type.
     */
    NPC('n', "npc", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The model primitive type.
     */
    MODEL('m', "model", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The top level interface primitive type.
     */
    TOPLEVELINTERFACE('F', "toplevelinterface", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The overlay interface primitive type.
     */
    OVERLAYINTERFACE('L', "overlayinterface", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The client interface primitive type.
     */
    CLIENTINTERFACE('\u00a9', "clientinterface", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The interface primitive type.
     */
    INTERFACE('a', "interface", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The component primitive type.
     */
    COMPONENT('I', "component", StackType.INT, -1, TypeSerializer.INT),

    /**
     * The long primitive type.
     */
    LONG('\u00cf', "long", StackType.LONG, 0L, TypeSerializer.LONG),

    /**
     * The boolean primitive type.
     */
    BOOLEAN('1', "boolean", StackType.INT, false, TypeSerializer.BOOLEAN),

    /**
     * The category primitive type.
     */
    CATEGORY('y', "category", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The namedobj primitive type.
     */
    NAMEDOBJ('O', "namedobj", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The object primitive type.
     */
    OBJ('o', "obj", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The inv primitive type.
     */
    INV('v', "inv", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The texture primitive type.
     */
    TEXTURE('x', "texture", StackType.INT, -1, TypeSerializer.BYTE),

    /**
     * The mapelement primitive type.
     */
    MAPELEMENT('\u00B5', "mapelement", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The graphic primitive type.
     */
    GRAPHIC('d', "graphic", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The struct primitive type.
     */
    STRUCT('J', "struct", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The loc primitive type.
     */
    LOC('l', "loc", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The colour primitive type.
     */
    COLOUR('C', "colour", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The identity kit primitive type.
     */
    IDK('K', "idkit", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The chat phrase primitive type.
     */
    CHATPHRASE('e', "chatphrase", StackType.INT, -1, TypeSerializer.SHORT),

    /**
     * The BAS primitive type.
     */
    BAS('\u20ac', "bas", StackType.INT, -1, TypeSerializer.SHORT);

    /**
     * The {@link PrimitiveType} by {@link #representation} look-up map.
     */
    private static final Map<String, PrimitiveType> lookupMap = Arrays.stream(values()).filter(PrimitiveType::isReferencable).collect(Collectors.toMap(PrimitiveType::getRepresentation, type -> type));

    /**
     * The code of this primitive type.
     */
    @Getter
    private final char code;

    /**
     * The primitive type textual representation.
     */
    @Getter
    private final String representation;

    /**
     * The stack which this type belongs or encodes to.
     */
    @Getter
    private final StackType stackType;

    /**
     * The default value of this type.
     */
    @Getter
    private final Object defaultValue;

    /**
     * The serializer the values of this type will use.
     */
    @Getter
    @SuppressWarnings("rawtypes")
    private final TypeSerializer serializer;

    /**
     * Casts this type and the specified other type to a highest type then compares if they are equals.
     *
     * @param other the other type that we want to compare.
     * @return <code>true</code> if they are equals otherwise <code>false</code>.
     */
    public boolean implicitEquals(PrimitiveType other) {
        return implicitCastToHigher(this) == implicitCastToHigher(other);
    }

    /**
     * Casts the specified {@link PrimitiveType type} to a higher type.
     *
     * @param type the type that we want to cast to a higher type.
     * @return the casted {@link PrimitiveType type} or the {@link PrimitiveType type} that was passed.
     */
    private PrimitiveType implicitCastToHigher(PrimitiveType type) {
        switch (type) {
            case BYTE:
            case SHORT:
            case TRIBYTE:
                return INT;
            default:
                return type;
        }
    }

    /**
     * Checks whether er or not the primitive type is a referencable type.
     *
     * @return <code>true</code> if it is otherwise <code>false</code>.
     */
    public boolean isReferencable() {
        switch (this) {
            case HOOK:
            case UNDEFINED:
            case BYTE:
            case SHORT:
            case TRIBYTE:
            case TYPE:
                return false;
            default:
                return representation != null;
        }
    }

    /**
     * Checks whether or not this {@link PrimitiveType type} is a declarable type. Which means that it can be used as
     * parameters, or local variable declarations.
     *
     * @return {@code true} if it is otherwise {@code false}.
     */
    public boolean isDeclarable() {
        return stackType != null;
    }

    /**
     * Checks whether or not the type can be used as an array.
     *
     * @return <code>true</code> if it does otherwise <code>false</code>.
     */
    public boolean isArrayable() {
        if (this == PrimitiveType.BOOLEAN) {
            return false;
        }
        return stackType == StackType.INT;
    }

    /**
     * Checks whether or not the primitive type belongs to config type.
     *
     * @return <code>true</code> if it does otherwise <code>false</code>.
     */
    public boolean isConfigType() {
        switch (this) {
            case SEQ:
            case STAT:
            case MAPAREA:
            case ENUM:
            case NPC:
            case CATEGORY:
            case NAMEDOBJ:
            case OBJ:
            case INV:
            case MAPELEMENT:
            case VAR:
            case VARBIT:
            case VARCINT:
            case VARCSTR:
            case STRUCT:
            case LOC:
            case PARAM:
            case FLO:
            case FLU:
            case SPOTANIM:
                return true;
            default:
                return false;
        }
    }


    /**
     * Checks whether or not this type is nullable.
     *
     * @return <code>true</code> if it is otherwise <code>false</code>.
     */
    public boolean isNullable() {
        if (stackType != StackType.INT) {
            return false;
        }
        switch (this) {
            case NULL:
            case PARAM:
                return false;
            default:
                return true;
        }
    }

    /**
     * Looks-up for the {@link PrimitiveType} with the textual representation.
     *
     * @param representation the textual representation of the {@link PrimitiveType}.
     * @return the {@link PrimitiveType} if found otherwise {@code null}.
     */
    public static PrimitiveType forRepresentation(String representation) {
        return lookupMap.get(representation);
    }

    /**
     * Returns the {@link PrimitiveType} with the specified {@code code}.
     *
     * @param code the code that we want it's associated primitive type.
     * @return the {@link PrimitiveType} if found otherwise {@code null}.
     */
    public static PrimitiveType forCode(char code) {
        for (var type : PrimitiveType.values()) {
            if (type.code == code) {
                return type;
            }
        }
        return null;
    }
}

File: runescript-commons/src/main/java/me/waliedyassen/runescript/type/primitive/PrimitiveType.kt
package me.waliedyassen.runescript.type.primitive

import me.waliedyassen.runescript.compiler.symbol.BasicSymbol
import me.waliedyassen.runescript.compiler.symbol.BasicSymbolLoader
import me.waliedyassen.runescript.compiler.symbol.ConfigSymbol
import me.waliedyassen.runescript.compiler.symbol.ConfigSymbolLoader
import me.waliedyassen.runescript.compiler.symbol.ConstantSymbol
import me.waliedyassen.runescript.compiler.symbol.ConstantSymbolLoader
import me.waliedyassen.runescript.compiler.symbol.Symbol
import me.waliedyassen.runescript.compiler.symbol.SymbolLoader
import me.waliedyassen.runescript.compiler.symbol.TypedSymbol
import me.waliedyassen.runescript.compiler.symbol.TypedSymbolLoader
import me.waliedyassen.runescript.type.Type
import me.waliedyassen.runescript.type.stack.StackType

open class PrimitiveType<T : Symbol>(
    override val code: Char,
    override val representation: String?,
    override val stackType: StackType?,
    override val defaultValue: Any?,
    val loader: SymbolLoader<T>?,
) : Type {


    object UNDEFINED : PrimitiveType<BasicSymbol>('\ufff0', "undefined", null, null, null)
    object HOOK : PrimitiveType<BasicSymbol>('\ufff1', "hook", null, null, null)
    object VOID : PrimitiveType<BasicSymbol>('\ufff2', "void", null, null, null)
    object CONSTANT : PrimitiveType<ConstantSymbol>('\ufff3', "constant", null, null, ConstantSymbolLoader)
    object TYPE : PrimitiveType<BasicSymbol>('\ufff6', "type", null, null, null)
    object PARAM : PrimitiveType<ConfigSymbol>('\uffd0', "param", null, null, ConfigSymbolLoader)
    object FLO : PrimitiveType<BasicSymbol>('\uffd1', "flo", null, null, null)
    object FLU : PrimitiveType<BasicSymbol>('\uffd2', "flu", null, null, null)
    object VARP : PrimitiveType<TypedSymbol>('\uffd3', "varp", null, null, TypedSymbolLoader)
    object VARBIT : PrimitiveType<BasicSymbol>('\uffd4', "varbit", null, null, BasicSymbolLoader)
    object VARC : PrimitiveType<TypedSymbol>('\uffd6', "varc", null, null, TypedSymbolLoader)
    object NULL : PrimitiveType<BasicSymbol>('\uffd7', null, null, null, null)
    object INT : PrimitiveType<BasicSymbol>('i', "int", StackType.INT, 0, null)
    object STRING : PrimitiveType<BasicSymbol>('s', "string", StackType.STRING, "", null)
    object SPOTANIM : PrimitiveType<BasicSymbol>('t', "spotanim", StackType.INT, -1, BasicSymbolLoader)
    object SEQ : PrimitiveType<BasicSymbol>('A', "seq", StackType.INT, -1, BasicSymbolLoader)
    object STAT : PrimitiveType<BasicSymbol>('S', "stat", StackType.INT, -1, BasicSymbolLoader)
    object SYNTH : PrimitiveType<BasicSymbol>('P', "synth", StackType.INT, -1, BasicSymbolLoader)
    object COORDGRID : PrimitiveType<BasicSymbol>('c', "coord", StackType.INT, -1, BasicSymbolLoader)
    object CHAR : PrimitiveType<BasicSymbol>('z', "char", StackType.INT, -1, BasicSymbolLoader)
    object FONTMETRICS : PrimitiveType<BasicSymbol>('f', "fontmetrics", StackType.INT, -1, BasicSymbolLoader)
    object MAPAREA : PrimitiveType<BasicSymbol>('`', "wma", StackType.INT, -1, BasicSymbolLoader)
    object ENUM : PrimitiveType<TypedSymbol>('g', "enum", StackType.INT, -1, TypedSymbolLoader)
    object NPC : PrimitiveType<BasicSymbol>('n', "npc", StackType.INT, -1, BasicSymbolLoader)
    object MODEL : PrimitiveType<BasicSymbol>('m', "model", StackType.INT, -1, BasicSymbolLoader)
    object TOPLEVELINTERFACE : PrimitiveType<BasicSymbol>('F', "toplevelinterface", StackType.INT, -1, BasicSymbolLoader)
    object OVERLAYINTERFACE : PrimitiveType<BasicSymbol>('L', "overlayinterface", StackType.INT, -1, BasicSymbolLoader)
    object CLIENTINTERFACE : PrimitiveType<BasicSymbol>('\u00a9', "clientinterface", StackType.INT, -1, BasicSymbolLoader)
    object INTERFACE : PrimitiveType<BasicSymbol>('a', "interface", StackType.INT, -1, BasicSymbolLoader)
    object COMPONENT : PrimitiveType<BasicSymbol>('I', "component", StackType.INT, -1, BasicSymbolLoader)
    object LONG : PrimitiveType<BasicSymbol>('\u00cf', "long", StackType.LONG, 0L, null)
    object BOOLEAN : PrimitiveType<BasicSymbol>('1', "boolean", StackType.INT, false, null)
    object CATEGORY : PrimitiveType<BasicSymbol>('y', "category", StackType.INT, -1, BasicSymbolLoader)
    object NAMEDOBJ : PrimitiveType<BasicSymbol>('O', "namedobj", StackType.INT, -1, BasicSymbolLoader)
    object OBJ : PrimitiveType<BasicSymbol>('o', "obj", StackType.INT, -1, BasicSymbolLoader)
    object INV : PrimitiveType<BasicSymbol>('v', "inv", StackType.INT, -1, BasicSymbolLoader)
    object TEXTURE : PrimitiveType<BasicSymbol>('x', "texture", StackType.INT, -1, BasicSymbolLoader)
    object MAPELEMENT : PrimitiveType<BasicSymbol>('\u00B5', "mapelement", StackType.INT, -1, BasicSymbolLoader)
    object GRAPHIC : PrimitiveType<BasicSymbol>('d', "graphic", StackType.INT, -1, BasicSymbolLoader)
    object STRUCT : PrimitiveType<BasicSymbol>('J', "struct", StackType.INT, -1, BasicSymbolLoader)
    object LOC : PrimitiveType<BasicSymbol>('l', "loc", StackType.INT, -1, BasicSymbolLoader)
    object COLOUR : PrimitiveType<BasicSymbol>('C', "colour", StackType.INT, -1, BasicSymbolLoader)
    object IDK : PrimitiveType<BasicSymbol>('K', "idkit", StackType.INT, -1, BasicSymbolLoader)
    object CHATPHRASE : PrimitiveType<BasicSymbol>('e', "chatphrase", StackType.INT, -1, BasicSymbolLoader)
    object BAS : PrimitiveType<BasicSymbol>('\u20ac', "bas", StackType.INT, -1, BasicSymbolLoader)
    object DBROW : PrimitiveType<BasicSymbol>('\u00D0', "dbrow", StackType.INT, -1, BasicSymbolLoader)
    object NEWVAR : PrimitiveType<BasicSymbol>('-', "newvar", StackType.INT, -1, BasicSymbolLoader)
    object NPC_UID : PrimitiveType<BasicSymbol>('u', "npc_uid", StackType.INT, -1, BasicSymbolLoader)
    object LOC_SHAPE : PrimitiveType<BasicSymbol>('H', "locshape", StackType.INT, -1, BasicSymbolLoader)

    override fun toString() = javaClass.simpleName

    val isReferencable: Boolean
        get() = when (this) {
            is TYPE -> false
            else -> representation != null
        }

    val isDeclarable: Boolean
        get() = stackType != null

    val isArrayable: Boolean
        get() = if (this == BOOLEAN) {
            false
        } else stackType == StackType.INT

    val isConfigType: Boolean
        get() = when (this) {
            is SEQ,
            is STAT,
            is MAPAREA,
            is ENUM,
            is NPC,
            is CATEGORY,
            is NAMEDOBJ,
            is OBJ,
            is INV,
            is MAPELEMENT,
            is VARP,
            is VARBIT,
            is VARC,
            is STRUCT,
            is LOC,
            is PARAM,
            is FLO,
            is FLU,
            is SPOTANIM -> true

            else -> false
        }

    val isNullable: Boolean
        get() = if (stackType != StackType.INT) {
            false
        } else when (this) {
            is NULL,
            is PARAM -> false

            else -> true
        }

    companion object {
        val values = listOf(
            UNDEFINED,
            HOOK,
            VOID,
            CONSTANT,
            TYPE,
            PARAM,
            FLO,
            FLU,
            VARP,
            VARBIT,
            VARC,
            NULL,
            INT,
            STRING,
            SPOTANIM,
            SEQ,
            STAT,
            SYNTH,
            COORDGRID,
            CHAR,
            FONTMETRICS,
            MAPAREA,
            ENUM,
            NPC,
            MODEL,
            TOPLEVELINTERFACE,
            OVERLAYINTERFACE,
            CLIENTINTERFACE,
            INTERFACE,
            COMPONENT,
            LONG,
            BOOLEAN,
            CATEGORY,
            NAMEDOBJ,
            OBJ,
            INV,
            TEXTURE,
            MAPELEMENT,
            GRAPHIC,
            STRUCT,
            LOC,
            COLOUR,
            IDK,
            CHATPHRASE,
            BAS,
            DBROW,
            NEWVAR,
            NPC_UID,
            LOC_SHAPE,
        )
        private val referencibleLookup = values
            .filter { it.isReferencable }
            .associateBy { it.representation!! }

        private val literalLookup = values
            .filter { it.representation != null }
            .associateBy { it.representation!! }

        @JvmStatic
        fun forRepresentation(representation: String): PrimitiveType<*>? {
            return referencibleLookup[representation]
        }
        @JvmStatic
        fun forLiteralOrNull(literal: String): PrimitiveType<*>? {
            return literalLookup[literal]
        }
        @JvmStatic
        fun forLiteral(literal: String): PrimitiveType<*> {
            return literalLookup[literal] ?: error("No type could be found for the literal '$literal'")
        }

    }
}
